gitlab/resource_gitlab_user_impersonation_token_test.go

```diff
@@ -0,0 +1,198 @@
+package gitlab
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	"github.com/xanzy/go-gitlab"
+)
+
+func TestAccGitlabUserImpersonationToken_basic(t *testing.T) {
+	var token gitlab.ImpersonationToken
+	rInt := acctest.RandInt()
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccGitlabUserImpersonationTokenDestroy,
+		Steps: []resource.TestStep{
+			// Create a user and impersonation token
+			// gitlab_user is already tested as part of `resource_gitlab_user`
+			{
+				Config: testAccGitlabUserImpersonationTokenConfig(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabUserImpersonationTokenExists("gitlab_user_impersonation_token.bar", &token),
+					testAccCheckGitlabUserImpersonationTokenAttributes(&token, &testAccCheckGitlabUserImpersonationTokenExpectedAttributes{
+						Name:    fmt.Sprintf("Token bar %d", rInt),
+						Active:  true,
+						Scopes:  []string{"api"},
+						Revoked: false,
+					}),
+				),
+			},
+		},
+	})
+}
+
+func TestAccGitlabUserImpersonationToken_withexpiration(t *testing.T) {
+	var token gitlab.ImpersonationToken
+	rInt := acctest.RandInt()
+	layout := "2006-01-02"
+	d, _ := time.Parse(layout, "2222-12-31")
+	iso_d := gitlab.ISOTime(d)
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccGitlabUserImpersonationTokenDestroy,
+		Steps: []resource.TestStep{
+			// Create a user and impersonation token with expiration
+			// gitlab_user is already tested as part of `resource_gitlab_user`
+			{
+				Config: testAccGitlabUserImpersonationTokenWithExpirationConfig(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabUserImpersonationTokenExists("gitlab_user_impersonation_token.will_expire", &token),
+					testAccCheckGitlabUserImpersonationTokenAttributes(&token, &testAccCheckGitlabUserImpersonationTokenExpectedAttributes{
+						Name:      fmt.Sprintf("Token will_expire %d", rInt),
+						Active:    true,
+						Scopes:    []string{"api", "read_user"},
+						Revoked:   false,
+						ExpiresAt: &iso_d,
+					}),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckGitlabUserImpersonationTokenExists(n string, token *gitlab.ImpersonationToken) resource.TestCheckFunc {
```

Comment:
There should also be an import test, e.g. https://github.com/terraform-providers/terraform-provider-vault/blob/master/vault/resource_token_test.go#L49-L75